packages/jest-core/src/cli/index.ts

```diff
@@ -222,36 +225,47 @@ const _run10000 = async (
   );
   performance.mark('jest/buildContextsAndHasteMaps:end');
 
-  globalConfig.watch || globalConfig.watchAll
-    ? await runWatch(
-        contexts,
-        configs,
-        hasDeprecationWarnings,
-        globalConfig,
-        outputStream,
-        hasteMapInstances,
-        filter,
-      )
-    : await runWithoutWatch(
-        globalConfig,
-        contexts,
-        outputStream,
-        onComplete,
-        changedFilesPromise,
-        filter,
-      );
+  if (globalConfig.watch || globalConfig.watchAll) {
+    await runWatch(
+      contexts,
+      configs,
+      warnForDeprecations,
+      globalConfig,
+      outputStream,
+      hasteMapInstances,
+      filter,
+    );
+    // If in watch mode, return the promise that will never resolve.
+    // If the watch mode is interrupted, watch should handle the process
+    // shutdown.
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    return new Promise(() => {});
+  } else {
+    let result: AggregatedResult;
+    await runWithoutWatch(
+      globalConfig,
+      contexts,
+      outputStream,
+      r => {
+        result = r;
+      },
+      changedFilesPromise,
+      filter,
+    );
+    return result!;
```

Comment:
Typing are correct here. For instance, if `bail` is set Jest will exit with returning any result, unfortunately:

https://github.com/facebook/jest/blob/285b40d5954fd81e07b02db92185fd317ba92578/packages/jest-core/src/TestScheduler.ts#L397-L419

Reply:
I'm not sure I understand what you mean. Afaik, whenever the promise was resolved, the `onComplete` was triggered with a result. If `bail` was set, it would exit with exit code 1, so the promise won't be resolved. 

I will document the `bail` behavior.

Reply:
Just trying to figure out how to make programmatic usage predictable. How to make it always return the result? If it is new wrapper function, perhaps it should not work in watch mode and ignore `bail`? So that result is always there?

Reply:
With regards to `bail`, i think we can leave that functionality in and later add a argument to pass an exit handler, as suggested in #11766. What do you think?

Removing watch functionality would be fine for my use case, I'm not using it from Stryker. On the other hand, it doesn't bother me 🤷‍♂️

Reply:
Yes, the issues with `bail` are out of scope.